### PR TITLE
Strategy to send all params in one request

### PR DESCRIPTION
**Pull request by rmaskell**
Right now each param was being sent individually per request, we have some urls where this will always cause a 400 because some params are missing and deeper api logic is not being tested. So i added `--strategy all_params_at_once` to send all of the params from the swagger in one request instead.

Also added a log to show the request body which i found useful.

### CODE REVIEW

**Comment by KissPeter** [comment]
Hi, 
Thanks for the PR It is the next step indeed.
Let me understand your code before merging...
Péter

File: apifuzzer/swagger_template_generator.py [hunk 1]
```diff
@@ -26,8 +27,8 @@ def normalize_url(url_in):
         # we will change back later
         return url_in.strip('/').replace('/', '+')
 
-    def process_api_resources(self):
-        self.logger.info('Start preparation')
+    def strategy_default():
+        self.logger.info('Start preparation with default strategy')
```
Comment (KissPeter) [inline apifuzzer/swagger_template_generator.py]: It should be the default strategy, no need to introduce new config parameter and duplicate the code

File: apifuzzer/base_template.py [hunk 2]
```diff
@@ -34,11 +35,26 @@ def __init__(self, name):
...
         """
 
+    def strategy_default(self, template):
+        for name, field in self.field_to_param.items():
+            if list(field):
+                template.append_fields([Container(name=name, fields=field)])
+
+    def strategy_all_params_at_once(self, template):
+        for name, field in self.field_to_param.items():
+            print(name)
```
Comment (KissPeter) [inline apifuzzer/base_template.py]: use logger instead of print.

Comment (rmaskell) [reply apifuzzer/base_template.py]: I think you're reviewing an old commit, this got removed.
I should squash my commits before opening a MR to make this easier for everyone.

Comment (KissPeter) [reply apifuzzer/base_template.py]: Yes I did :-)

File: apifuzzer/swagger_template_generator.py (hunk 1 above)
Comment (rmaskell) [reply apifuzzer/swagger_template_generator.py]: fixed in latest commits

**Comment by KissPeter** [comment]
Hi Richard,

It seems some  alignment is necessary.
I've made a test with this input:
  _"paths": {
    "/multiple_path_param/{param1}/{param2}": {
      "get": {
        "parameters": [
          {
            "name": "param1",
            "in": "path",
            "required": true,
            "type": "number",
            "format": "double"
          },
          {
            "name": "param2",
            "in": "path",
            "required": true,
            "type": "number",
            "format": "double"
          }
        ]
      }
    }
  }_


and got queries like this:
**127.0.0.1 - - [08/Feb/2020 21:26:39] "GET /multiple_path_param/667.5/param2_http://127.0.0.1:5000/multiple_path_param/_param1/ HTTP/1.1" 404 -**
I assume the expand_path_variables function should be reviewed

Without your modification only one parameter is used:
**127.0.0.1 - - [08/Feb/2020 21:29:07] "GET /multiple_path_param/param1/5!]2yNone HTTP/1.1" 404 -**

For multiple query string parameters your extension works great:
**127.0.0.1 - - [08/Feb/2020 21:29:57] "GET /multiple_query_paramts?queryparam0=asd&queryparam1=Dr% HTTP/1.1" 404 -**

Would you fix it or shall I do it?

Thanks

**Comment by rmaskell** [comment]
> Hi Richard,
> 
> It seems some alignment is necessary.
> I've made a test with this input:
> _"paths": { "/multiple_path_param/{param1}/{param2}": { "get": { "parameters": [ { "name": "param1", "in": "path", "required": true, "type": "number", "format": "double" }, { "name": "param2", "in": "path", "required": true, "type": "number", "format": "double" } ] } } }_
> 
> and got queries like this:
> **127.0.0.1 - - [08/Feb/2020 21:26:39] "GET /multiple_path_param/667.5/param2_[http://127.0.0.1:5000/multiple_path_param/_param1/�](http://127.0.0.1:5000/multiple_path_param/_param1/%10) HTTP/1.1" 404 -**
> I assume the expand_path_variables function should be reviewed
> 
> Without your modification only one parameter is used:
> **127.0.0.1 - - [08/Feb/2020 21:29:07] "GET /multiple_path_param/param1/5!]2yNone HTTP/1.1" 404 -**
> 
> For multiple query string parameters your extension works great:
> **127.0.0.1 - - [08/Feb/2020 21:29:57] "GET /multiple_query_paramts?queryparam0=asd&queryparam1=�Dr% HTTP/1.1" 404 -**
> 
> Would you fix it or shall I do it?
> 
> Thanks

Hi Peter,

thank you for catching this, i'll take a look now.

**Comment by rmaskell** [comment]
> Hi Richard,
> 
> It seems some alignment is necessary.
> I've made a test with this input:
> _"paths": { "/multiple_path_param/{param1}/{param2}": { "get": { "parameters": [ { "name": "param1", "in": "path", "required": true, "type": "number", "format": "double" }, { "name": "param2", "in": "path", "required": true, "type": "number", "format": "double" } ] } } }_
> 
> and got queries like this:
> **127.0.0.1 - - [08/Feb/2020 21:26:39] "GET /multiple_path_param/667.5/param2_[http://127.0.0.1:5000/multiple_path_param/_param1/�](http://127.0.0.1:5000/multiple_path_param/_param1/%10) HTTP/1.1" 404 -**
> I assume the expand_path_variables function should be reviewed
> 
> Without your modification only one parameter is used:
> **127.0.0.1 - - [08/Feb/2020 21:29:07] "GET /multiple_path_param/param1/5!]2yNone HTTP/1.1" 404 -**
> 
> For multiple query string parameters your extension works great:
> **127.0.0.1 - - [08/Feb/2020 21:29:57] "GET /multiple_query_paramts?queryparam0=asd&queryparam1=�Dr% HTTP/1.1" 404 -**
> 
> Would you fix it or shall I do it?
> 
> Thanks

image

image


should be fixed now :) @KissPeter

**Comment by KissPeter** [comment]
looks great thank you for fixing it

**Pull request merged**